plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs

```diff
@@ -401,14 +401,20 @@ mkTx lookups txc = mkSomeTx [SomeLookupsAndConstraints lookups txc]
 
 -- | Each transaction output should contain a minimum amount of Ada (this is a
 -- restriction on the real Cardano network).
-adjustUnbalancedTx :: Params -> UnbalancedTx -> Either Tx.ToCardanoError UnbalancedTx
-adjustUnbalancedTx params = mapMOf (tx . Tx.outputs . traverse) adjustTxOut
+adjustUnbalancedTx :: Params -> UnbalancedTx -> Either Tx.ToCardanoError ([Ada.Ada], UnbalancedTx)
+adjustUnbalancedTx params utx =
+    let (acc, res) = mapAccumLOf (tx . Tx.outputs . traverse) step (Right []) utx
+    in (flip (,) res) <$> acc
   where
-    adjustTxOut :: TxOut -> Either Tx.ToCardanoError TxOut
+    step acc txOut = case (acc, adjustTxOut txOut) of
+        (Left _, _)                       -> (acc, txOut) -- acc is an error, do nothing
+        (Right _, Left e)                 -> (Left e, txOut) -- failed somewhere, return error as acc
+        (Right acc', Right (ada, txOut')) -> (Right $ ada:acc', txOut')
+    adjustTxOut :: TxOut -> Either Tx.ToCardanoError (Ada.Ada, TxOut)
```

Comment:
If you let this return `Either Tx.ToCardanoError ([Ada.Ada], TxOut)` you have 2 nested applicative effects (`Either Tx.ToCardanoError` and `(,) [Ada.Ada]`), which you could use in a traversal using `Compose`. Then you won't need `step`.

Reply:
This worked even better than I thought, trying it out in vscode I got:
```haskell
adjustUnbalancedTx params = alaf Compose (tx . Tx.outputs . traverse) adjustTxOut
```

Reply:
Oh, indeed, looks cool and works fine! Thanks!

But lacks some readability as non-trivial lens code as usual.